File: BOJ_백준/15900 나무 탈출.py
import sys
sys.setrecursionlimit(10**6)
input=sys.stdin.readline

def dfs(n, depth):
    global cnt
    visited[n] = True

    if n != 1 and len(tree[n]) == 1 :
        if depth % 2 == 1 :
            cnt += 1
            return
    
    for i in tree[n]:
        if not visited[i] :
            dfs(i, depth + 1)

n=int(input())
tree=[[] for _ in range(n + 1)]
visited=[False] * (n + 1)
cnt=0

for _ in range(n - 1):
    a, b = map(int,input().split())
    tree[a].append(b)
    tree[b].append(a)

dfs(1,0)
print('Yes' if cnt % 2 == 1 else 'No')

File: BOJ_백준/15900 나무 탈출.cpp
#include<iostream>
#include<algorithm>
#include<vector>

using namespace std;

int N, cnt;
bool visited[500001];
vector<int> tree[500001];

void dfs(int n, int depth)
{
	// leaf node
	if (n != 1 && tree[n].size() == 1)
	{
		if (depth % 2 == 1) cnt++;
		return;
	}

	for (int i = 0; i < tree[n].size(); i++)
	{
		if (visited[n] == false)
		{
			visited[n] = true;
			dfs(tree[n][i], depth + 1);
			visited[n] = false;
		}
	}
}

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cin >> N;
	int a, b;
	for (int i = 0; i < N - 1; i++)
	{
		cin >> a >> b;
		tree[a].push_back(b);
		tree[b].push_back(a);
	}

	dfs(1, 0);
	if (cnt % 2 == 1) cout << "Yes";
	else cout << "No";
	return 0;
}